### PR TITLE
Migrate to thiserror and remove dep on derive-error

### PR DESCRIPTION
**Pull request by delta1**
## Description
All error types migrated to use `thiserror` crate. Dependency on `derive-error` removed. 

## Motivation and Context
#1718 motivates the migration of error types to `thiserror` crate in place of `derive-error`.
Also bumped `thiserror` to version `1.0.20` across the repo. 
In places where previous error messages were defined for `derive-error` I have used those. 
Where not previously defined, I have used generic descriptions of the errors - please advise if these are not suitable?

## How Has This Been Tested?
* [X] cargo test 
* [X] circle ci pipeline
* [X] manual sanity test

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [X] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
* [X] I'm merging against the `development` branch.
* [X] I ran `cargo-fmt --all` before pushing.
* [X] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/base_node/chain_metadata_service/error.rs [hunk 1]
```diff
@@ -21,20 +21,23 @@
...
-    /// Peer did not send any chain metadata
+    #[error("Failed to decode chain metadata")]
+    DecodeError(#[from] DecodeError),
+    #[error("Peer did not send any chain metadata")]
     NoChainMetadata,
-    LivenessError(LivenessError),
-    CommsInterfaceError(CommsInterfaceError),
-    MessageError(MessageError),
-    /// Failed to publish `ChainMetadataEvent`
+    #[error("Liveness Error")]
```
Comment (sdbondi) [inline base_layer/core/src/base_node/chain_metadata_service/error.rs]: Few other places where the actual error is not being output.
 
```suggestion
    #[error("Liveness Error: {0}")]
```

File: base_layer/core/src/base_node/service/error.rs [hunk 2]
```diff
@@ -21,15 +21,17 @@
...
-use derive_error::Error;
 use tari_comms_dht::outbound::DhtOutboundError;
+use thiserror::Error;
 
 #[derive(Debug, Error)]
 pub enum BaseNodeServiceError {
-    CommsInterfaceError(CommsInterfaceError),
-    DhtOutboundError(DhtOutboundError),
-    #[error(msg_embedded, no_from, non_std)]
+    #[error("Comms interface error")]
```
Comment (sdbondi) [inline base_layer/core/src/base_node/service/error.rs]:
```suggestion
    #[error("Comms interface error: {0}")]
```

same for most of the rest

File: base_layer/core/src/mempool/error.rs [hunk 3]
```diff
@@ -30,21 +30,26 @@ use crate::{
...
 #[derive(Debug, Error)]
 pub enum MempoolError {
-    UnconfirmedPoolError(UnconfirmedPoolError),
-    OrphanPoolError(OrphanPoolError),
-    PendingPoolError(PendingPoolError),
-    ReorgPoolError(ReorgPoolError),
-    TransactionError(TransactionError),
-    ChainStorageError(ChainStorageError),
-    /// The Blockchain height is undefined
+    #[error("Unconfirmed pool error")]
```
Comment (sdbondi) [inline base_layer/core/src/mempool/error.rs]:
```suggestion
    #[error("Unconfirmed pool error: {0}")]
```

**Review by sdbondi — COMMENTED** [review]
LGTM 🍸  - think that this PR should just make sure the info contained in each variant on the enum is output otherwise we could get a few useless errors like "Database Error" and "Message error"

```rust
 #[error("Message error: {0}")] // Add the `{0}`
    MessageError(#[from] MessageError),
```

File: base_layer/core/src/base_node/chain_metadata_service/error.rs (hunk 1 above)
Comment (delta1) [reply base_layer/core/src/base_node/chain_metadata_service/error.rs]: thanks @sdbondi - I thought that was only necessary for the "String" types, and #[from] took care of the others? 

How do I know which errors need that interpolation - is it all of them?

Comment (sdbondi) [reply base_layer/core/src/base_node/chain_metadata_service/error.rs]: `#[from]` generates a [`From`](https://doc.rust-lang.org/std/convert/trait.From.html) implementation to make it easy to convert say a `LivenessError` to a `ChainMetadataError`.

What `#[error("Liveness Error: {0}")]` is doing is calling the [`Display`](https://doc.rust-lang.org/std/fmt/trait.Display.html) implementation on `LivenessError` (which itself is implemented using `thiserror` and so has a `Display` implementation.

Comment (sdbondi) [reply base_layer/core/src/base_node/chain_metadata_service/error.rs]: So I think most of them should have the "inner error" output - basically it depends on the meaning of the error, which is hard to evaluate in bulk, but we definitely want to err on the side of displaying the info vs hiding it.

**Comment by delta1** [comment]
Perfect thanks for the feedback @sdbondi I'll get that done

**Review by sdbondi — DISMISSED** [review]
:shipit:

**Comment by delta1** [comment]
@sdbondi @mikethetike this PR is updated after merging all the current stuff from dev and fighting with git for a while

**Review by stringhandler — APPROVED** [review]

**Pull request merged**